### ISSUE TITLE
Add-ons: Fix/add automated tests for Calypso & Signup

### ISSUE THREAD

**Issue by chriskmnds** [Automattic/wp-calypso#65409]
There's a new admin page and a step in Signup for purchasing add-ons. The latter (and potentially both) is subject to an [ExPlat test](https://github.com/Automattic/wp-calypso/pull/65343). Once that's decided, let's review the code ([admin page](https://github.com/Automattic/wp-calypso/pull/64289) & [signup](https://github.com/Automattic/wp-calypso/pull/64451)) and determine the level of testing we need for this.

Clearly, the new Signup step may cause breakage once the flag is enabled after the test (if successful). Might be good to plan any updates before the day.

**Comment by chriskmnds** [comment]
CC @worldomonation

**Comment by worldomonation** [comment]
@chriskmnds Thanks for the heads up! 

I agree, let's prepare something so that it can be merged together with the PR that enables the screens in question. 
How do you see me fitting into the overall project?
1. advise how to go about making changes so that it doesn't break Pre-Release Tests before or after the feature is enabled.
2. work with whoever is implementing the feature on the same branch to add/modify specs as needed.
3. work in parallel with the feature branch to add/modify the specs as needed, and merge together with the feature PR.

**Comment by chriskmnds** [comment]
Thanks @worldomonation. Either way is fine. I am not sure what sense of urgency we have here (@gmovr should know more). It might be the case we send all users to "treatment" variation for a brief period of time once decided, which should give some flexibility to work on specs before full release? (obviously not sure)

For signup, there was one bug we easily missed while coding, which may be worth adding some coverage for. We were not able to go back from checkout and add more add-ons to the cart (https://github.com/Automattic/wp-calypso/issues/65341). This was eventually resolved, but I suspect it will be easy to reintroduce without noticing again. Another likely addition will be to have the add-ons step reflect what's in the cart (it currently doesn't), which may be another candidate for test coverage.

**Comment by worldomonation** [comment]
I went and checked how the Add-ons page behaves on both `wpcalypso.wordpress.com` and production.

![image](https://user-images.githubusercontent.com/6549265/178602433-59688620-7680-49c3-8d8b-34ebf8eea0e1.png)

Confirmed that e2e test accounts do not see this A/B test. 

As for the paths that would be suitable to test - we currently have a spec to cover the signup process as a new user. I feel that we can add the Add-ons selection to this flow to avoid the need to create yet another spec, as the flows for a hypothetical standalone Free with Add-ons spec would be essentially the same thing as the existing Free spec.

**Comment by worldomonation** [comment]
@chriskmnds With the Plans revert, is this still relevant?

**Comment by chriskmnds** [comment]
> @chriskmnds With the Plans revert, is this still relevant?

I believe so. The add-ons haven't been reverted, at least not yet. I believe they are supposed to stay in both signup and dashboard - but CC @gmovr to confirm status

**Comment by worldomonation** [comment]
@chriskmnds Checking in a few months later. According to p1662049928110199-slack-C03D472D64C, it appears that add-ons have been canned for now. In that case, I think it would make sense to close this task for now.

**Comment by chriskmnds** [comment]
> According to p1662049928110199-slack-C03D472D64C, it appears that add-ons have been canned for now.

Hey @worldomonation ! I gather add-ons are no longer part of Signup. I am not sure if we should close this issue as we still have add-ons in Calypso that may need the addition of tests (whatever form, we'd probably need to revisit and decide). I guess you can remove the task from the board concerning adding Signup E2E tests though (what you are referring to?). Can you please maybe confirm the status of add-ons in Signup with someone from @Automattic/martech first (@southp maybe)?

**Comment by southp** [comment]
@chriskmnds @worldomonation The add-on flow won't be back anytime soon and the experiment should be deprecated. I've created a task for our team here: https://github.com/Automattic/wp-calypso/issues/68330. For this issue, I'd suggest to either opening a new one explicitly for the admin page, or making it clear that this issue will only be about the admin page. How does that sound?

**Comment by chriskmnds** [comment]
> @chriskmnds @worldomonation The add-on flow won't be back anytime soon and the experiment should be deprecated. I've created a task for our team here: #68330. For this issue, I'd suggest to either opening a new one explicitly for the admin page, or making it clear that this issue will only be about the admin page. How does that sound?

Thanks @southp! Sounds great (well and kinda sad that add-ons flow didn't make it). Either way is fine by me too. The Signup E2E tests were the priority here obviously. So it looks like a kind of lower priority cleanup task now to see what/if anything we need for the admin section. Something like:

_**Title:** Add-ons: unit/automated tests for admin section.
**Description:** As of https://github.com/Automattic/wp-calypso/pull/64289 , a new admin section under upgrades renders a list of add-ons for users to purchase. Let's examine what/if any opportunities we missed to introduce some automated/unit tests there (for hooks, components, etc.). Some previous chat in https://github.com/Automattic/wp-calypso/issues/65409_

> I've created a task for our team here: https://github.com/Automattic/wp-calypso/issues/68330.

Nice! 🦸

**Comment by worldomonation** [comment]
> I guess you can remove the task from the board concerning adding Signup E2E tests though (what you are referring to?).

I guess I'm trying to figure out if this is something that's worth doing, and if so the priority of it.

> > I am not sure if we should close this issue as we still have add-ons in Calypso that may need the addition of tests (whatever form, we'd probably need to revisit and decide).

My impression was that Add-ons were not rolled out to production.

If the above statement is false and thus Add-ons were rolled out to customers, what should happen when they go to renew in about 9 months' time? From what I gather in your comments this is the scenario you're referring to and proposing E2E coverage for, is this correct?

**Comment by southp** [comment]
@worldomonation The add-ons did roll out to customers; just not in the signup flow. It can be accessed under the Upgrades tab:
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/1842898/193231919-ed298960-839a-4406-b163-c342d23544f4.png">

Thus, my understanding to @chriskmnds 's comment is to cover key actions in this page like we did for the plans page, i.e. a user can successfully purchase an add-on in this page :)

**Comment by chriskmnds** [comment]
> Thus, my understanding to @chriskmnds 's comment is to cover key actions in this page like we did for the plans page, i.e. a user can successfully purchase an add-on in this page :)

Hey @worldomonation! Just to second what @southp said above :-) 

There are also a few "finer" things that we could potentially test in some form, but probably less critical i.e. the cards become inactive and show a "purchased" or "included in your plan" text depending on the respective action taken (user purchased the add-on or purchased a plan with that feature).

**Comment by chriskmnds** [comment]
> I guess I'm trying to figure out if this is something that's worth doing, and if so the priority of it.

@worldomonation Thanks for looking into this! 

My impression is if add-ons will remain in Calypso admin, then good chance they'll become integral to the bigger picture, with potentially more add-ons being introduced ("storage" maybe next, according to that Slack thread?). I think some e2e testing would suffice vs unit testing any individual parts here if we proceed. :-)
